lib/DxcSupport/Unicode.cpp

```diff
@@ -38,62 +40,78 @@ int MultiByteToWideChar(uint32_t /*CodePage*/, uint32_t /*dwFlags*/,
     // Add 1 for the null-terminating character.
     ++cbMultiByte;
   }
-  // if zero is given as the destination size, this function should
+  // If zero is given as the destination size, this function should
   // return the required size (including the null-terminating character).
+  // This is the behavior of mbstowcs when the target is null.
   if (cchWideChar == 0) {
-    wchar_t *tempStr = (wchar_t *)malloc(cbMultiByte * sizeof(wchar_t));
-    size_t requiredSize = mbstowcs(tempStr, lpMultiByteStr, cbMultiByte);
-    free(tempStr);
-    if (requiredSize == (size_t)cbMultiByte) return requiredSize;
-    return requiredSize + 1;
-  }
-
-  if (cchWideChar < cbMultiByte) {
+    lpWideCharStr = nullptr;
+  } else if (cchWideChar < cbMultiByte) {
     SetLastError(ERROR_INSUFFICIENT_BUFFER);
     return 0;
   }
 
-  size_t rv = mbstowcs(lpWideCharStr, lpMultiByteStr, cbMultiByte);
+  size_t rv;
+  const char *locale = CPToLocale(CodePage);
+  locale = setlocale(LC_ALL, locale);
+  if (lpMultiByteStr[cbMultiByte] != '\0') {
```

Comment:
You could avoid malloc/free by doing this ?

```cpp
char c = lpMultiByteStr[cbMultiByte];

// Make sure source is null-terminated.
if (c != '\0')
  lpMultiByteStr[cbMultiByte] = '\0';

rv = mbstowcs(lpWideCharStr, lpMultiByteStr, cchWideChar);

// If we modified the source, undo our change.
if (c != '\0')
  lpMultiByteStr[cbMultiByte] = c;
```

also applies to WideCharToMultiByte

Reply:
I considered this, but sadly, lpMultiByteStr is a const pointer.

Reply:
Ah right!